File: src/main/java/dark/leech/text/plugin/js/api/JSList.java
package dark.leech.text.plugin.js.api;

import java.util.ArrayList;

/**
 * Special list type for vBook plugin compatibility using Rhino. Returned by Html.map() and supports
 * JavaScript array-like access. Simple Java class - public methods automatically exposed to
 * JavaScript.
 */
public class JSList extends ArrayList<Object> {

    public JSList() {
        super();
    }

    public JSList(int initialCapacity) {
        super(initialCapacity);
    }

    /** Add item to list. */
    @Override
    public boolean add(Object item) {
        return super.add(item);
    }

    /** Get item at index. */
    public Object get(int index) {
        if (index >= 0 && index < size()) {
            return super.get(index);
        }
        return null;
    }

    /** Convert to JavaScript array. */
    public Object[] toArray() {
        return super.toArray();
    }

    /** ForEach iteration - passes element and index to callback. */
    public void forEach(Object callback) {
        if (callback instanceof org.mozilla.javascript.Function) {
            org.mozilla.javascript.Function func = (org.mozilla.javascript.Function) callback;
            org.mozilla.javascript.Context ctx = org.mozilla.javascript.Context.getCurrentContext();
            org.mozilla.javascript.Scriptable scope = func.getParentScope();

            for (int i = 0; i < size(); i++) {
                Object item = get(i);
                func.call(ctx, scope, scope, new Object[] {item, i});
            }
        }
    }

    /** Map - transform each element, returns new JSList. */
    public JSList map(Object callback) {
        JSList result = new JSList(size());
        if (callback instanceof org.mozilla.javascript.Function) {
            org.mozilla.javascript.Function func = (org.mozilla.javascript.Function) callback;
            org.mozilla.javascript.Context ctx = org.mozilla.javascript.Context.getCurrentContext();
            org.mozilla.javascript.Scriptable scope = func.getParentScope();

            for (int i = 0; i < size(); i++) {
                Object item = get(i);
                Object mapped = func.call(ctx, scope, scope, new Object[] {item, i});
                result.add(mapped);
            }
        }
        return result;
    }

    /** Filter elements by predicate. */
    public JSList filter(Object predicate) {
        JSList result = new JSList();
        if (predicate instanceof org.mozilla.javascript.Function) {
            org.mozilla.javascript.Function func = (org.mozilla.javascript.Function) predicate;
            org.mozilla.javascript.Context ctx = org.mozilla.javascript.Context.getCurrentContext();
            org.mozilla.javascript.Scriptable scope = func.getParentScope();

            for (int i = 0; i < size(); i++) {
                Object item = get(i);
                Object testResult = func.call(ctx, scope, scope, new Object[] {item, i});
                if (testResult instanceof Boolean && (Boolean) testResult) {
                    result.add(item);
                }
            }
        }
        return result;
    }

    /** Find first element matching predicate. */
    public Object find(Object predicate) {
        if (predicate instanceof org.mozilla.javascript.Function) {
            org.mozilla.javascript.Function func = (org.mozilla.javascript.Function) predicate;
            org.mozilla.javascript.Context ctx = org.mozilla.javascript.Context.getCurrentContext();
            org.mozilla.javascript.Scriptable scope = func.getParentScope();

            for (int i = 0; i < size(); i++) {
                Object item = get(i);
                Object testResult = func.call(ctx, scope, scope, new Object[] {item, i});
                if (testResult instanceof Boolean && (Boolean) testResult) {
                    return item;
                }
            }
        }
        return null;
    }

    /** Join elements with separator. */
    public String join(String separator) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < size(); i++) {
            if (i > 0) {
                sb.append(separator);
            }
            Object item = get(i);
            if (item != null) {
                sb.append(org.mozilla.javascript.Context.toString(item));
            }
        }
        return sb.toString();
    }

    /** Check if list contains element. */
    public boolean contains(Object element) {
        return super.contains(element);
    }

    /** Get first element. */
    public Object first() {
        return isEmpty() ? null : get(0);
    }

    /** Get last element. */
    public Object last() {
        return isEmpty() ? null : get(size() - 1);
    }

    /** Check if list is empty. */
    public boolean isEmpty() {
        return super.isEmpty();
    }
}

File: engine/src/commonMain/kotlin/dev/haipham22/leechtext/plugin/js/api/JSList.kt
package dev.haipham22.leechtext.plugin.js.api

/**
 * List type cho vBook compat (port từ JSList.java; P5.2c: composition thay extends
 * ArrayList — ArrayList final trên Kotlin/Native). Trả về từ JSElements.map(),
 * hỗ trợ array-like access từ JavaScript (get/length qua NativeJavaObject methods).
 */
@Suppress("TooManyFunctions") // API surface cho plugin JS — số hàm mirror đúng API vBook, không tách
class JSList(
    initialCapacity: Int = 16,
) {
    private val items: MutableList<Any?> = ArrayList(initialCapacity)

    val size: Int get() = items.size

    val length: Int get() = items.size

    fun isEmpty(): Boolean = items.isEmpty()

    fun add(element: Any?): Boolean = items.add(element)

    operator fun get(index: Int): Any? = if (index in 0 until size) items[index] else null

    fun toList(): List<Any?> = items.toList()

    fun forEach(callback: Any?) {
        if (JsValueBridge.isFunction(callback)) {
            for (i in 0 until size) {
                JsValueBridge.call(callback, listOf(get(i), i))
            }
        }
    }

    fun map(callback: Any?): JSList {
        val result = JSList(size)
        if (JsValueBridge.isFunction(callback)) {
            for (i in 0 until size) {
                result.add(JsValueBridge.call(callback, listOf(get(i), i)))
            }
        }
        return result
    }

    fun filter(predicate: Any?): JSList {
        val result = JSList()
        if (JsValueBridge.isFunction(predicate)) {
            for (i in 0 until size) {
                val item = get(i)
                val testResult = JsValueBridge.call(predicate, listOf(item, i))
                if (testResult is Boolean && testResult) result.add(item)
            }
        }
        return result
    }

    fun find(predicate: Any?): Any? {
        if (JsValueBridge.isFunction(predicate)) {
            for (i in 0 until size) {
                val item = get(i)
                val testResult = JsValueBridge.call(predicate, listOf(item, i))
                if (testResult is Boolean && testResult) return item
            }
        }
        return null
    }

    fun join(separator: String): String {
        val sb = StringBuilder()
        for (i in 0 until size) {
            if (i > 0) sb.append(separator)
            get(i)?.let { sb.append(JsValueBridge.toJsString(it) ?: "null") }
        }
        return sb.toString()
    }

    fun first(): Any? = if (isEmpty()) null else get(0)

    fun last(): Any? = if (isEmpty()) null else get(size - 1)
}
